pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -0,0 +1,182 @@
+// would be in turn borrowed from underlying account. Once event is
+// processed, allocation would be returned.
+// If budget is exhausted, allocations would block until more is available
+// or until an optional timeout expires.
+// NB: if we are competing with other consumers in the budget pool, then
+// we won't see that memory is released.
+type feedBudget struct {
+	mu struct {
+		syncutil.Mutex
+		// Bound account that provides budget with resource.
+		memBudget *mon.BoundAccount
+		// Whenever we pass oversized message this would go to true to prevent
+		// other attempts to acquire more memory until the event goes through.
+		oversize bool
+		// If true, budget was released and no more allocations could take place.
+		closed bool
+	}
+	// Channel to notify that memory was returned to the budget.
+	replenishC chan interface{}
+
+	metrics *Metrics
+}
+
+func (f *feedBudget) init(budget *mon.BoundAccount, metrics *Metrics) {
+	f.mu.memBudget = budget
+	if budget != nil {
+		f.replenishC = make(chan interface{}, 1)
+		f.metrics = metrics
+	}
+}
+
+// Get allocates amount from budget.
+// If budget has less memory than requested it could either wait, if we have any
+// in flight allocations, or it could send a single message through regardless
+// of its size. In latter case, memory is not requested from underlying budget,
+// but we won't be able to do another Get until this message is returned. To
+// track such events, feedBudget uses a counter metric.
+//
+// timeout argument indicates how long we could wait for the budget to replenish
+// if there is not enough budget available. Budget is replenished by Return
+// calls.
+// timeout value of 0 means wait indefinitely.
+// Whenever we use timeout, we return its wait chan to caller, so it could be
+// used subsequently by the chained wait for other resources thus keeping total
+// timeout in check without adding to it on every op. If we didn't initiate
+// a wait or if timeout was 0 then wait channel will be nil.
+// Returned allocation has its use counter set to 1.
+func (f *feedBudget) Get(
+	ctx context.Context, amount int64, timeout time.Duration, stoppedC <-chan struct{},
+) (sharedBudgetAllocation, <-chan time.Time, error) {
+	// Function to allocate from budget or if there's nothing there borrow from
+	// emergency ether.
+	tryAlloc := func() (sharedBudgetAllocation, error) {
+		f.mu.Lock()
+		defer f.mu.Unlock()
+		if f.mu.closed {
+			return sharedBudgetAllocation{}, errors.New("budget is already released")
+		}
+		if f.mu.oversize {
+			return sharedBudgetAllocation{}, errors.New("failed to allocate mem budget")
+		}
+		err := f.mu.memBudget.Grow(ctx, amount)
+		if err == nil {
+			return sharedBudgetAllocation{size: amount, refCount: 1}, nil
+		}
+		// If we didn't use any budget yet and not currently sending an oversized
+		// event then we can send one.
+		if f.mu.memBudget.Used() == 0 {
+			f.metrics.RangeFeedOverBudgetEvents.Inc(1)
+			f.mu.oversize = true
```

Comment:
This should claim the maximum amount possible from the budget. If the system is memory-constrained, we don't want to use a huge amount of memory here that isn't accounted for at all. In that case, wouldn't that also mean that the `oversize` field becomes redundant?

Reply:
That was my original intent, but it can't easily be done without modifying BytesMonitor and related types as it hides most of the details and provides usage info and grow requests. I'll see how it could be done in a least intrusive way.

Reply:
Have you had a look at how SQL does this? I imagine they'll have similar issues there, since I believe row values can be arbitrarily large.

Reply:
We never do such tricks in SQL land. We either bail straight away or try to free memory, reuse previous rows, swap out to disk before bailing. Maybe we can have a separate monitor that we would allocate against which will have no limit and will act as a monitor only?

Reply:
So one thing I didn't consider is that the size of these events will necessarily be bounded by `kv.raft.command.max_size`, which defaults to 64MB. As long as the budget is never set smaller that that, I don't think we need this single-large-event handling at all.

Reply:
That sounds plausable, but I think we want an oversize for the cases when we are starving because other feeds or sql or kv depleted budget. Override will still allow us to trickle events. What do you think?

Reply:
I think it's fine to drop events in the case -- if we're out of budget, we're out of budget.

But there is a broader question here about fairness and QoS. For example, we should prioritize system rangefeeds over user rangefeeds. I think we'll do that by having a separate reserved budget for system rangefeeds or something, right? For user rangefeeds, it might be useful to give separate budgets or priorities to individual changefeeds, but that would have to tie in with changefeed configuration somehow and I think we can keep that out of scope for now?

Reply:
Isn't that tricky in general to achieve:
- shared budget (not only with feeds, but with other components)
- fairness so everyone could advance in case of starvation
- isolate "bad actors'
- don't reserve memory upfront to avoid idle feeds wasting memory
Quota pool might work here, but only within feeds themselves. Maybe we need to understand a bigger picture, what do we expect in case of starvation?